### DISCUSSION TITLE
Not able to retrive the created entity: 401 unauthorised

### DISCUSSION THREAD

**Discussion by radhay-samagra** [Sunbird-RC/community#220] [Issues]
I have registered one schema and added an entity for the same with the following response

## Schema:
```
{
	"$schema": "http://json-schema.org/draft-07/schema",
	"type": "object",
	"properties": {
		"board-hp": {
			"$ref": "#/definitions/board-hp"
		}
	},
	"required": ["board-hp"],
	"title": "board-hp",
	"definitions": {
		"board-hp": {
			"$id": "#/properties/board-hp",
			"type": "object",
			"title": "samarth-hp schema",
			"required": ["name", "phoneNumber", "email"],
			"uniqueIndexFields": ["phoneNumber", "email"],
			"properties": {
				"name": {"type": "string"},
                "phoneNumber": {"type": "string"},
                "email": {"type": "string"}
			}
		}
	},
	"_osConfig": {
		"ownershipAttributes": [
			{
                "email": "/email",
                "mobile": "/phoneNumber",
                "userId": "/phoneNumber"
              }
		],
		"inviteRoles": ["anonymous"],
		"enableLogin": true
	}
}
```

### API Request to create entity
```
curl --location --request POST \
--header 'content-type: application/json' \
--data-raw '{
    "name": "test-hp",
    "phoneNumber": "1234567890",
    "email": "test@example.com"
}'\
'http://localhost:8081/api/v1/board-hp/invite'
```
>  where board-hp is the name of the entity
### Response
```
{
    "id": "sunbird-rc.registry.invite",
    "ver": "1.0",
    "ets": 1651062273207,
    "params": {
        "resmsgid": "",
        "msgid": "8a7054fe-810b-4099-bfed-1859e065dc6e",
        "err": "",
        "status": "SUCCESSFUL",
        "errmsg": ""
    },
    "responseCode": "OK",
    "result": {
        "samarth-hp": {
            "osid": "1-72d3a4e2-4f37-4ec9-867d-be7748018885"
        }
    }
```

Now I wanted to retrieve my entity via API  ```GET /api/v1/{entity-type}/{id} HTTP/1.1```

I made the following request after successfully authenticating as an entity and generated an access token:
``` access-token: eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJWek9OWUU0TEdObHh1WS1leHdCbmk0Z0tTWmNUaXVaSmlKdF9QeGpyQ01BIn0.eyJleHAiOjE2NTEwNjQyMTAsImlhdCI6MTY1MTA2MzYxMCwianRpIjoiMmI4MzQ0ZTctMjI0My00NmFlLWIwMGUtNWI5MDJiNDFhYTRkIiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo4MDgwL2F1dGgvcmVhbG1zL3N1bmJpcmQtcmMiLCJhdWQiOiJhY2NvdW50Iiwic3ViIjoiZGFiMDg1MjQtNmY2NC00ZTlkLWIzNmEtNjJhMjU3YmEyOGI3IiwidHlwIjoiQmVhcmVyIiwiYXpwIjoicmVnaXN0cnktZnJvbnRlbmQiLCJzZXNzaW9uX3N0YXRlIjoiYWQzMDg0OGYtMzI5Yy00N2U5LTgzODYtYWZlNjU4YzkwZTQyIiwiYWNyIjoiMSIsImFsbG93ZWQtb3JpZ2lucyI6WyJodHRwczovL2xvY2FsaG9zdDo0MjAyIiwiaHR0cDovL2xvY2FsaG9zdDo0MjAyIiwiaHR0cHM6Ly9sb2NhbGhvc3Q6NDIwMCIsImh0dHBzOi8vbmRlYXIueGl2LmluIiwiaHR0cDovL2xvY2FsaG9zdDo0MjAwIiwiaHR0cDovL25kZWFyLnhpdi5pbiIsImh0dHA6Ly8yMC4xOTguNjQuMTI4Il0sInJlYWxtX2FjY2VzcyI6eyJyb2xlcyI6WyJvZmZsaW5lX2FjY2VzcyIsInVtYV9hdXRob3JpemF0aW9uIiwiZGVmYXVsdC1yb2xlcy1uZGVhciJdfSwicmVzb3VyY2VfYWNjZXNzIjp7ImFjY291bnQiOnsicm9sZXMiOlsibWFuYWdlLWFjY291bnQiLCJtYW5hZ2UtYWNjb3VudC1saW5rcyIsInZpZXctcHJvZmlsZSJdfX0sInNjb3BlIjoicHJvZmlsZSBlbWFpbCIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwicHJlZmVycmVkX3VzZXJuYW1lIjoiMTIzNDU2Nzg5MCIsImVudGl0eSI6WyJzYW1hcnRoLWhwIl0sImVtYWlsIjoidGVzdEBleGFtcGxlLmNvbSJ9.Fw8BV_QnUedyTF-uK3cGpBCLc0NCmugyfMSkvMY_fdMe43MzhZgTMbKUt4nr05x19Vl9iCRI55YRMhAJ82NX5q-femAxtrdj8FuXEpxak4VBEXH3SmsfD95X5LrokcMfoHpL-qeZtQy3tF9LwUVgEH7KOP0JiQYfimjXNdUgI48nta_rikIqp2PsgN6TVCWzkOKXduZBtdCnMt6kqXod8luAjckfxwKukqgzs4I2hIagkisiL_pupLadDkX0WON0yB53qChDsHCZb2FFc-qL0XOGOP9nAm4mKBrrLlg4jmuEhKWG-5myUS1QSrJBDQktB5cDMpBvBQB3Fy1ml7YC9w```

```
curl --location \
	--request GET \
	--header 'content-type: application/json' \
	--header 'authorization: bearer {access-token}' \
	'http://localhost:8081/api/v1/board-hp/1-d5408ef9-f6b6-47f3-a334-ba85eaaf141a'

```
I am getting 401  as an HTTP response and I am also not able to find any significant logs by the service to debug.

**Comment by tejash-jl** [comment]
@radhay-samagra 
I think you got the keycloak token by calling `http://localhost:8080/auth/realms/sunbird-rc/protocol/openid-connect/token`,
instead, can you use **`http://keycloak:8080/auth/realms/sunbird-rc/protocol/openid-connect/token`** and also add the 
```
127.0.0.1	keycloak
```
to `/etc/hosts`

**Reply by radhay-samagra** [reply]
@tejash-jl  
I have tried forwarding my 127.0.0.1 to keycloak and generated a token too:
```
curl --location --request POST 'http://keycloak:8080/auth/realms/sunbird-rc/protocol/openid-connect/token' \
--header 'content-type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=registry-frontend' \
--data-urlencode 'username=1234567890' \
--data-urlencode 'password=abcd@123' \
--data-urlencode 'grant_type=password'
```

I tried using the generated token for retrieving my entity using the same api used above but I am still getting 403 Forbidden and below log:
```
registry_1      | 16:34:19.809 [http-nio-8081-exec-3] INFO  d.s.registry.helper.RegistryHelper:261 - Read Api: shard id: 1 for label: 1-c6718644-69ae-49a4-87df-592b2f61787d
```
Noot sure why this is occuring. I was able to use this API when using the example teacher and student schema mentioned in the docs.

**Reply by gamemaker1** [reply]
@radhay-samagra could you please share the entire logs from the moment you make the request? Thank you

**Reply by radhay-samagra** [reply]
@gamemaker1 
This is the only trace I got from the moment I authenticated and made a request to retrieve API
```
claim-ms_1      | 2022-04-28 17:26:25.208 DEBUG 1 --- [l-1 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Pool stats (total=10, active=0, idle=10, waiting=0)
claim-ms_1      | 2022-04-28 17:26:25.208 DEBUG 1 --- [l-1 housekeeper] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Fill pool skipped, pool is at sufficient level.
claim-ms_1      | 2022-04-28 17:26:28.675 DEBUG 1 --- [alina-utility-1] org.apache.catalina.session.ManagerBase  : Start expire sessions StandardManager at 1651166788674 sessioncount 0
claim-ms_1      | 2022-04-28 17:26:28.675 DEBUG 1 --- [alina-utility-1] org.apache.catalina.session.ManagerBase  : End expire sessions StandardManager processingTime 1 expired sessions: 0
registry_1      | 17:26:39.851 [http-nio-8081-exec-1] INFO  d.s.registry.helper.RegistryHelper:261 - Read Api: shard id: 1 for label: 1-c6718644-69ae-49a4-87df-592b2f61787d
```

**Reply by gamemaker1** [reply]
I don't see any errors here; could you please paste the response body and headers when you get the 403?

**Reply by radhay-samagra** [reply]
@gamemaker1 
There is no response body for the API, just ```403: Forbidden``` as HTTP status and not many relevant logs too. I mentioned this earlier. I am not able to trace where it's going wrong.

**Reply by gamemaker1** [reply]
What about the headers?

**Reply by pramodkvarma** [reply]
Folks, Discord server for all Sunbird projects are up and running. [Here is the invite for Discord Sunbird-RC channel](https://discord.gg/Q5mvw2mGC8). Please join and have live discussions there. We are now fully off slack. 

@surendrasinghs fyi

**Reply by radhay-samagra** [reply]
@gamemaker1 
I have mentioned the headers in the curl requests shared above. I am using just the shared ones only.

**Reply by gamemaker1** [reply]
I meant the headers that you receive in response, sometimes there is an error included in them.

**Reply by radhay-samagra** [reply]
@gamemaker1 
PFA

![image](https://user-images.githubusercontent.com/90022986/165817695-0371127a-c3a1-4646-bc25-9d70fb718d5b.png)

**Reply by gamemaker1** [reply]
I'm sorry, there does not seem to be any error here either.

One more question though: has this registry been setup with consent enabled?

**Reply by radhay-samagra** [reply]
No consent is disabled. I cross checked on the keycloak admin dashboard too

**Reply by gamemaker1** [reply]
Then I'm not sure what the error could be; I'll try setting up a registry myself tomorrow and debug the problem.

**Reply by radhay-samagra** [reply]
I can help you with the boilerplate and the configs I have used for replicating the issue on discord.

**Reply by gamemaker1** [reply]
Yes please, that would be very helpful

**Comment by tejash-jl** [comment]
@radhay-samagra,
403 issue occurs if authorized user tries to access other entities,
can you check if the osid  you are using in the GET API is the same that you received in the invite API response,
Also you can also use this GET API: `http://localhost:8081/api/v1/board-hp` without osid

**Reply by radhay-samagra** [reply]
@tejash-jl There was some issue with the regustry-frontend client. I created a new client kn keycloak and I was able to fetch the entities now. Also forwarding 127.0.0.1 to keycloak was necessary apparently. I thought it would not be that important. My issue is now resolved. Thanks.

**Reply by gamemaker1** [reply]
Good to hear that. Could you please share the config for the registry frontend client as well as the new client you created, maybe we can find out what the problem was from there (just export realm configuration in keycloak, and share the resultant json file). Thank you
